server/jetstream_cluster.go

```diff
@@ -1241,21 +1241,67 @@ func (js *jetStream) applyStreamEntries(mset *Stream, ce *CommittedEntry) (bool,
 	return didSnap, nil
 }
 
+// Returns the PeerInfo for all replicas of a raft node. This is different than node.Peers()
+// and is used for external facing advisories.
+func (s *Server) replicas(node RaftNode) []*PeerInfo {
+	now := time.Now()
+	var replicas []*PeerInfo
+	for _, rp := range node.Peers() {
+		pi := &PeerInfo{Name: s.serverNameForNode(rp.ID), Current: rp.Current, Active: now.Sub(rp.Last)}
+		replicas = append(replicas, pi)
+	}
+	return replicas
+}
+
 func (js *jetStream) processStreamLeaderChange(mset *Stream, sa *streamAssignment, isLeader bool) {
+	js.mu.Lock()
+	s, account, err := js.srv, sa.Client.Account, sa.err
+	client, reply := sa.Client, sa.Reply
+	hasResponded := sa.responded
+	sa.responded = true
+	js.mu.Unlock()
+
+	stream := mset.Name()
+
 	if isLeader {
-		js.srv.Noticef("JetStream cluster new stream leader for '%s > %s'", sa.Client.Account, mset.Name())
+		s.Noticef("JetStream cluster new stream leader for '%s > %s'", sa.Client.Account, stream)
+		if node := mset.raftNode(); node != nil {
+			s.publishAdvisory(mset.account(), JSAdvisoryStreamLeaderElectedPre+"."+stream, &JSStreamLeaderElectedAdvisory{
+				TypedEvent: TypedEvent{
+					Type: JSStreamLeaderElectedAdvisoryType,
+					ID:   nuid.Next(),
+					Time: time.Now().UTC(),
+				},
+				Stream:   stream,
+				State:    mset.State(),
+				Leader:   s.serverNameForNode(node.GroupLeader()),
+				Replicas: s.replicas(node),
+			})
+		}
+	} else {
+		// We are stepping down. Make sure if we are doing so because we have lost quorum that
+		// we send the appropriate advisories.
+		if node := mset.raftNode(); node != nil && !node.Quorum() {
+			s.Warnf("JetStream cluster stream '%s > %s' has lost quorum, stalled.", sa.Client.Account, stream)
+			s.publishAdvisory(mset.account(), JSAdvisoryStreamQuorumLostPre+"."+stream, &JSStreamQuorumLostAdvisory{
+				TypedEvent: TypedEvent{
+					Type: JSStreamQuorumLostAdvisoryType,
+					ID:   nuid.Next(),
+					Time: time.Now().UTC(),
+				},
+				Stream:   stream,
+				State:    mset.State(),
```

Comment:
without consumer state i wonder if its worth having this really? I am not sure what I'll do with this info if I had it, what did you have in mind as a usecase?

Reply:
What would you do with state here?

Reply:
I can't think of anything I'd do with it :)